gcc/rust/util/rust-token-converter.cc

```diff
@@ -0,0 +1,586 @@
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-lex.h"
+#include "rust-token-converter.h"
+#include "libproc_macro/proc_macro.h"
+
+namespace Rust {
+
+static void
+pop_group (std::vector<ProcMacro::TokenStream> &streams,
+	   ProcMacro::Delimiter delim)
+{
+  auto g = ProcMacro::Group::make_group (streams.back (), delim);
+  streams.pop_back ();
+  auto tt = ProcMacro::TokenTree::make_tokentree (g);
+
+  streams.back ().push (tt);
+}
+
+static void
+dispatch_float_literals (ProcMacro::TokenStream &ts,
+			 const const_TokenPtr &token)
+{
+  std::string::size_type sz;
+  auto str = token->as_string ();
+  switch (token->get_type_hint ())
+    {
+      case CORETYPE_F32: {
+	auto value = std::stof (str, &sz);
+	ts.push (ProcMacro::TokenTree::make_tokentree (
+	  ProcMacro::Literal::make_f32 (value, sz == str.length ())));
```

Comment:
so floats are suffixed when the size is equal to the token's string's size? So if we have the token `15f32`, which is suffixed (unless I'm misunderstanding?): we'll `stdof` it, get `15` and a size of 2 in `sz` (?). So the expression `sz == str.length()` will be false and our f32 literal won't be suffixed?

Reply:
Yes, there is an inversion of the condition here (and everywhere this mechanism is applied), good catch.

Reply:
Fixed in [0182ef1](https://github.com/Rust-GCC/gccrs/pull/2170/commits/0182ef1edc9d0e5431cd2b5e5b552754886f997d)